### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by emreyalvac**

### CODE REVIEW

**Comment by emreyalvac** [comment]
@dorfsmay 

Hi,

Can u review please :)

**Comment by emreyalvac** [comment]
@dorfsmay @y2kappa @Bobbbay

Hi,

Can u review please :)

File: src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+use std::fmt::{Display, Formatter};
+
+#[derive(Debug)]
+pub enum UrlParseError {
+    NoPath,
+    NotHttps,
+    Parser(url::ParseError),
+}
```
Comment (Bobbbay) [inline src/errors.rs:8]: Do you really need this, or does `url`'s `ParseError` cover everything? It seems to me that it does: https://docs.rs/url/latest/url/enum.ParseError.html#variants

File: src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    let uri = Url::parse(uri.as_str());
+
+    let uri = match uri {
+        Ok(res) => res,
+        Err(err) => return Err(UrlParseError::Parser(err)),
+    };
+
+    if uri.scheme() != "https" {
+        return Err(UrlParseError::NotHttps);
+    }
```
Comment (Bobbbay) [inline src/utils.rs:14]: What's the reasoning behind mandating `https`? If Firebase has a problem with `http`, it should probably be reported from their side, not manually from the library.

File: src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+use std::fmt::{Display, Formatter};
+
+#[derive(Debug)]
+pub enum UrlParseError {
```
Comment (Bobbbay) [inline src/errors.rs:4]: You should create a type alias to `Result<T, UrlParseError>` and use it instead, e.g. in `src/utils.rs`' `check_uri` function.

File: src/errors.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        match self {
+            UrlParseError::NoPath => write!(f, "URL path is missing."),
+            UrlParseError::NotHttps => write!(f, "The URL protocol should be https."),
+            UrlParseError::Parser(e) => write!(f, "Error while parsing the URL: {}", e),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum RequestError {
```
Comment (Bobbbay) [inline src/errors.rs:21]: The same comment regarding a custom `Result` type goes for here.

File: src/lib.rs [hunk 5]
```diff
@@ -1,564 +1,249 @@
-/*
-Firebase REST API
-*/
```
Comment (Bobbbay) [inline src/lib.rs:3]: This top-level crate documentation was a great idea. Maybe not in this PR, but eventually I highly recommend writing doc comments for everything, including top-level module documentation.

**Review by Bobbbay — COMMENTED** [review]
In case you're still looking for a code review, I've added a few comments here.

**Pull request merged**